### PR TITLE
Actions stable release

### PR DESCRIPTION
**Pull request by bholmesdev**
## Changes

This makes Actions stable for v4.15.
- Remove `experimental` flag 
- Remove deprecated `getActionProps()` API
- Add check for whether actions are used before mounting routes and middleware
- Add check for a define action symbol before registering an action function. Maybe overkill, but this 100% ensures the change is non-breaking for users that _happen_ to have an `actions/index.ts` file that isn't actions related.

## Testing

Remove experimental flag and `getActionProps()` from tests

## Docs

https://github.com/withastro/docs/pull/8980

### CODE REVIEW

File: packages/astro/src/actions/index.ts
Comment (bholmesdev) [reply packages/astro/src/actions/index.ts:29]: It's not an integration. It's in core 🙃

File: packages/astro/src/integrations/hooks.ts
Comment (bholmesdev) [reply packages/astro/src/integrations/hooks.ts:133]: Good point! Let's move to a top-level import

File: packages/astro/src/actions/index.ts
Comment (bholmesdev) [reply packages/astro/src/actions/index.ts:29]: I have realized this check isn't full-proof though. To make it backwards compatible, we should check if the `server` export is present before booting up actions

Comment (florian-lefebvre) [reply packages/astro/src/actions/index.ts:29]: I know it's in core but we are actually inserting an integration in the astro config, that's what i meant. like instead of checking in each hook, check as early as possible

Comment (bholmesdev) [reply packages/astro/src/actions/index.ts:29]: Ah, perhaps we could. The issue is that `astro:server:setup` hook, which we _do_ want to mount to detect when actions are added in your project. If we conditionally add the whole integration, we can't check for this

Comment (florian-lefebvre) [reply packages/astro/src/actions/index.ts:29]: Ah I didn't know, current approach seems go then

**Comment by bholmesdev** [comment]
@florian-lefebvre I revisited your comment on simplifying the integration. I realized we could do so! I split things up so we have `vitePluginActions()` to define the `astro:actions`, module, and the `astroIntegrationActionsRouterHandler()` to _only_ inject routes when actions are used. I moved that file watcher to a vite plugin. Should solve a failing test in the process too 👍

**Review by matthewp — APPROVED** [review]

File: .changeset/spicy-suits-explode.md
Comment (sarah11918) [inline .changeset/spicy-suits-explode.md:7]: No notes on the included text! Just giving you permission in case you wanted to expand here a bit with an example, maybe like how the experimental changeset would have looked (updated and correct of course) or copying from the deleted experimental config reference content below.

I will not un-approve if this gets longer! At your discretion, @bholmesdev !

**Review by sarah11918 — APPROVED** [review]
Just giving Ben more permission to hype it up in the changeset as he sees fit! Errors nicely updated as well! Approving for docs!

File: packages/astro/src/actions/integration.ts
Comment (florian-lefebvre) [inline packages/astro/src/actions/integration.ts:20]: I think this should occur in config:done instead since it could be updated during any integration config:setup

**Review by bluwy — APPROVED** [review]

File: packages/astro/src/actions/plugins.ts
Comment (ematipico) [inline packages/astro/src/actions/plugins.ts:68]: When do we change the value of `isActionsUsed`? If line 65 sets the value to `true`, then here we `false` and `server.restart()` is never called again.

File: packages/astro/src/actions/runtime/virtual/get-action.ts
Comment (ematipico) [inline packages/astro/src/actions/runtime/virtual/get-action.ts:14]: Do we have a case where `server` is `undefined`? If so, I think it should be an error instead

File: packages/astro/src/actions/utils.ts
Comment (ematipico) [inline packages/astro/src/actions/utils.ts:33]: Can we add a comment that explains what this function does? The logic is really weird and I can't understand it

File: packages/astro/src/actions/utils.ts
Comment (ematipico) [inline packages/astro/src/actions/utils.ts:65]: Why do we use this pattern to look for an action file? This pattern is really similar to `middleware.js` - where users can have `middleware.js` or `middleware/index.js`, and more -  and in that case we use the `moduleLoader` to retrieve the file.

It's not a blocker, but I think we should agree on **one way** to do things internally. 

Using the module loader from vite would allow us also to remove the use of `es-module-lexer`

**Review by ematipico — COMMENTED** [review]
I have some reservations on how we load the actions file, because it adds _yet another way to do things_. 

I don't want to block the PR, but I believe we should address it.

File: packages/astro/src/actions/runtime/virtual/get-action.ts
Comment (bholmesdev) [reply packages/astro/src/actions/runtime/virtual/get-action.ts:14]: Ah, I added that from an older prototype, but it's no longer possible to get `undefined`. I can add an internal error to flag when `actionLookup` is unexpectedly undefined or some other type

File: .changeset/spicy-suits-explode.md
Comment (bholmesdev) [reply .changeset/spicy-suits-explode.md:7]: Haha, appreciate that. I'll copy a snippet from the 4.15 release blog post

File: packages/astro/src/actions/integration.ts
Comment (bholmesdev) [reply packages/astro/src/actions/integration.ts:20]: Oh good point. I guess it would error _after_ we attempt to add middleware, but I guess that's fine

File: packages/astro/src/actions/utils.ts
Comment (bholmesdev) [reply packages/astro/src/actions/utils.ts]: Good call. I renamed this to `isActionsFilePresent` and added an in-line comment on why we use lexer.

File: packages/astro/src/actions/utils.ts
Comment (bholmesdev) [reply packages/astro/src/actions/utils.ts:72]: Discussed on discord. We discovered that we can't rely on Vite to check whether actions are used, since we need that information _before_ vite starts to inject routes. We're tabling this for now to find a better solution.

**Review by sarah11918 — APPROVED** [review]
One little changeset nit, otherwise both changesets are approved by docs!

**Review by florian-lefebvre — APPROVED** [review]

**Comment by kenedyolisi** [comment]
Actions is one of the many features I love in astro. I couldn't wait for it to be stable and the provided guide is so extensive. I looove it. Thanks to @bholmesdev for tremendous effort on this and other contributors too for this feature.

**Comment by bholmesdev** [comment]
Ah, you made my day! Thanks @kenedyolisi

**Pull request merged**
